sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/ext/WalletService.kt

```diff
@@ -35,7 +34,43 @@ suspend inline fun retryUpTo(
             if (failedAttempts > retries) {
                 throw exceptionWrapper(t)
             }
-            val duration = (initialDelayMillis.toDouble() * Math.pow(2.0, failedAttempts.toDouble() - 1)).toLong()
+            val duration = (initialDelayMillis.toDouble() * 2.0.pow(failedAttempts.toDouble() - 1)).toLong()
+            Twig.warn(t) { "Retrying ($failedAttempts/$retries) in ${duration}s..." }
+            delay(duration)
+        }
+    }
+}
+
+/**
+ * Execute the given block and if it fails, retry up to [retries] more times. If none of the
+ * retries succeed, then leave the block execution unfinished and continue.
+ *
+ * @param retries the number of times to retry the block after the first attempt fails.
+ * @param exceptionWrapper a function that can wrap the final failure to add more useful information
+ *  * or context. Default behavior is to just return the final exception.
+ * @param initialDelayMillis the initial amount of time to wait before the first retry.
+ * @param block the code to execute, which will be wrapped in a try/catch and retried whenever an
+ * exception is thrown up to [retries] attempts.
+ */
+suspend inline fun retryUpToAndContinue(
+    retries: Int,
+    exceptionWrapper: (Throwable) -> Throwable = { it },
+    initialDelayMillis: Long = 500L,
+    block: (Int) -> Unit
+) {
+    var failedAttempts = 0
+    while (failedAttempts < retries) {
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            block(failedAttempts)
+            return
+        } catch (t: Throwable) {
```

Comment:
I know that in Java, it's considered bad form to catch `Throwable` because that includes for example `InterruptedException` and `ThreadDeath`, which interferes with thread termination.

Reply:
Yes, totally agree. But I'd keep this implementation, as we really need to catch any kind of exception and report it to the caller. That's why we have suppression for the lint checks here `@Suppress("TooGenericExceptionCaught")`. We could switch this older pattern to Kotlin's [`runCatching`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/run-catching.html), but that does the same inside:
```
/**
 * Calls the specified function [block] and returns its encapsulated result if invocation was successful,
 * catching any [Throwable] exception that was thrown from the [block] function execution and encapsulating it as a failure.
 */
@InlineOnly
@SinceKotlin("1.3")
public inline fun <R> runCatching(block: () -> R): Result<R> {
    return try {
        Result.success(block())
    } catch (e: Throwable) {
        Result.failure(e)
    }
}
```